libraries/ESP8266LLMNR/ESP8266LLMNR.cpp

```diff
@@ -0,0 +1,284 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * Reference:
+ * https://tools.ietf.org/html/rfc4795 (LLMNR)
+ * https://tools.ietf.org/html/rfc1035 (DNS)
+ */
+
+#include <debug.h>
+#include <functional>
+#include <ESP8266LLMNR.h>
+#include <ESP8266WiFi.h>
+#include <WiFiUdp.h>
+
+extern "C" {
+#include <user_interface.h>
+}
+
+#include <lwip/udp.h>
+#include <lwip/igmp.h>
+#include <include/UdpContext.h>
+
+//#define LLMNR_DEBUG
+
+#define BIT(x) (1 << (x))
+
+#define FLAGS_QR                BIT(15)
+#define FLAGS_OP_SHIFT          11
+#define FLAGS_OP_MASK           0xf
+#define FLAGS_C                 BIT(10)
+#define FLAGS_TC                BIT(9)
+#define FLAGS_T                 BIT(8)
+#define FLAGS_RCODE_SHIFT       0
+#define FLAGS_RCODE_MASK        0xf
+
+#define _conn_read16() (((uint16_t)_conn->read() << 8) | _conn->read())
+#define _conn_read8() _conn->read()
+#define _conn_readS(b, l) _conn->read((b), (l));
+
+static const IPAddress LLMNR_MULTICAST_ADDR(224, 0, 0, 252);
+static const int LLMNR_MULTICAST_TTL = 1;
+static const int LLMNR_PORT = 5355;
+
+LLMNRResponder::LLMNRResponder() :
+    _conn(0) {
+}
+
+LLMNRResponder::~LLMNRResponder() {
+    if (_conn) {
+        _conn->unref();
+    }
+}
+
+bool LLMNRResponder::begin(const char* hostname) {
+    // Max length for a single label in DNS
+    if (strlen(hostname) > 63) {
+        return false;
+    }
+
+    _hostname = hostname;
+    _hostname.toLowerCase();
+
+    WiFi.onStationModeGotIP([this](const WiFiEventStationModeGotIP& event){
+        _restart();
+    });
+
+    WiFi.onStationModeDisconnected([this](const WiFiEventStationModeDisconnected& event) {
+        _restart();
+    });
+
+    return _restart();
+}
+
+void LLMNRResponder::notify_ap_change() {
+    _restart();
+}
+
+bool LLMNRResponder::_restart() {
+    if (_conn) {
+        _conn->unref();
+        _conn = 0;
+    }
+
+    ip_addr_t multicast_addr;
+    multicast_addr.addr = (uint32_t)LLMNR_MULTICAST_ADDR;
+
+    if (igmp_joingroup(IP_ADDR_ANY, &multicast_addr) != ERR_OK)
+        return false;
+
+    _conn = new UdpContext;
+    _conn->ref();
+
+    if (!_conn->listen(*IP_ADDR_ANY, LLMNR_PORT))
```

Comment:
It's a minor thing, but in some other places you put braces around one-line if bodies. Suggest doing the same thing everywhere.